File: tests/xcalablemp/global-view/bcast/C/bcast.c
#include <xmp.h>
#include <stdio.h>
#include <stdlib.h>
#pragma xmp nodes p(*)
static const int N=1000;
int i,j,aa[1000],a;
double bb[1000],b;
float cc[1000],c;
int procs, result=0;

int main(void)
{
  procs = xmp_num_nodes();
  a = xmp_node_num();
  b = (double)a;
  c = (float)a;
  for(i=0;i<N;i++){
    aa[i] = a+i;
    bb[i] = (double)(a+i);
    cc[i] = (float)(a+i);
  }

#pragma xmp bcast (a)
#pragma xmp bcast (b) 
#pragma xmp bcast (c) 
#pragma xmp bcast (aa) 
#pragma xmp bcast (bb) 
#pragma xmp bcast (cc)
  j = 1;
  if(a != j)         result = -1;
  if(b != (double)j) result = -1;
  if(c != (float)j)  result = -1;
  for(i=0;i<N;i++){
    if(aa[i] != j+i)           result = -1;
    if(bb[i] != (double)(j+i)) result = -1;
    if(cc[i] != (float)(j+i))  result = -1;
   }

#pragma xmp reduction(+:result)
#pragma xmp task on p(1)
  {
    if(result == 0){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }
  
  return 0;
}    

File: tests/XMP/global-view/bcast/F/bcast.f90
program test

  !$xmp nodes p(*)

  real :: c(100,100)
  real :: d(100,100)

  call sub(c, d)

  !$xmp task on p(1)
  write(*,*) "PASS"
  !$xmp end task

contains

  subroutine sub(c, d)

    real :: a(100,100)
    real, allocatable :: b(:,:)
    real :: c(100,100)
    real :: d(:,:)

    integer :: result = 0

    allocate (b(100,100))

    !$xmp task on p(1)
    a = 1.
    b = 2.
    c = 3.
    d = 4.
    !$xmp end task

! ---------------------------------------------

    !$xmp bcast (a)

    do j = 1, 100
       do i = 1, 100
          if (a(i,j) /= 1.) result = -1
       end do
    end do

    !$xmp reduction(+:result)

    !$xmp task on p(1)
    if (result /= 0) then
       write(*,*) "ERROR"
       call exit(1)
    endif
    !$xmp end task

! ---------------------------------------------

    !$xmp bcast (b)

    do j = 1, 100
       do i = 1, 100
          if (b(i,j) /= 2.) result = -1
       end do
    end do

    !$xmp reduction(+:result)

    !$xmp task on p(1)
    if (result /= 0) then
       write(*,*) "ERROR"
       call exit(1)
    endif
    !$xmp end task

! ---------------------------------------------

    !$xmp bcast (c)

    do j = 1, 100
       do i = 1, 100
          if (c(i,j) /= 3.) result = -1
       end do
    end do

    !$xmp reduction(+:result)

    !$xmp task on p(1)
    if (result /= 0) then
       write(*,*) "ERROR"
       call exit(1)
    endif
    !$xmp end task

! ---------------------------------------------

    !$xmp bcast (d)

    do j = 1, 100
       do i = 1, 100
          if (d(i,j) /= 4.) result = -1
       end do
    end do

    !$xmp reduction(+:result)

    !$xmp task on p(1)
    if (result /= 0) then
       write(*,*) "ERROR"
       call exit(1)
    endif
    !$xmp end task

  end program

end program test

